### PR TITLE
New module: support Alibaba Cloud ECS (cloud/alicloud/alicloud_instance)

### PR DESCRIPTION
**Pull request by xiaozhu36**
SUMMARY

Added a new module alicloud_instance.py that can manage alicloud ecs instance.

COMPONENT NAME

lib/ansible/module_utils/alicloud_ecs.py
lib/ansible/modules/cloud/alicloud/alicloud_instance.py
lib/ansible/modules/cloud/alicloud/alicloud_instance_facts.py

ANSIBLE VERSION

2.4.0

### CODE REVIEW

**Review by ryansb — CHANGES_REQUESTED** [review]
Please rename this module. Users call modules within a flat namespace, so the module file should be `alicloud_ecs` to avoid confusion with an ECS provided by a different cloud.

**Comment by xiaozhu36** [comment]
Hi @ryansb 

Thank you for your feedback. I have renamed the module, and please check it, thanks.

Regards,
Guimin

**Comment by xiaozhu36** [comment]
Hi @ryansb @gundalow and Ansible Team:

I'm an Alibaba Cloud development engineer. I have finished the first PR of Alicloud new modules, and this module can be used to create Alicloud ECS instance and get its facts.

At present, there are many users want to use our module to manage their infrastructure and deploy their applications in it. So my team and I earnestly hope that you can review and integrate the PR into devel branch as soon as possible, and then released it in the next version.

According to your PR requirements, I will push other alicloud modules after the first PR being merged. If the module has any defect, please contact me and I would resolve and improve it timely, thanks.

Looking forward to your feedback, thanks.

Regards,
Guimin He

**Comment by dharmabumstead** [comment]
Please do not merge until I've had a chance to do an edit pass on the doc updates. Thanks.

**Comment by xiaozhu36** [comment]
Hi @dharmabumstead 

Thank you for your contribution. Looking forward to the doc update and your good news. Thanks a lot.

Regards,
Guimin He

**Comment by xiaozhu36** [comment]
Hi @dharmabumstead 

Ours many users want to use our module to manage their infrastructure and deploy their applications. It has been 13 days since my first PR, but it seems that my PR still is not reviewed and integrated into devel branch. Does it have some work I can do so that the progress can be speeded? If so, please tell me and I will try my best to do it, thanks a lot.

Regards,
He Guimin

**Comment by gundalow** [comment]
Hi,
This hasn't made the cut off date for 2.4, so this can not be merged till we've branched.

Once all the issues are resolved it will be merged and released into Ansible 2.5.

**Comment by xiaozhu36** [comment]
HI @gundalow 

Thank you for your feedback. Looking forward to your good news. Thanks a lot.

Regards,
He Guimin

File: docs/docsite/rst/guide_alicloud.rst
Comment (gundalow) [inline docs/docsite/rst/guide_alicloud.rst]: I think this line can be removed.

File: docs/docsite/rst/guide_alicloud.rst
Comment (gundalow) [inline docs/docsite/rst/guide_alicloud.rst]: Do you what version is required?

Will the version in the different OS package managers be new enough?

File: docs/docsite/rst/guide_alicloud.rst
Comment (gundalow) [inline docs/docsite/rst/guide_alicloud.rst]: Capital L in Alicloud-related.

File: docs/docsite/rst/guide_alicloud.rst
Comment (gundalow) [inline docs/docsite/rst/guide_alicloud.rst]: Might be good to link to the Ansible Vault documentation here, so people don't have passwords visible in their vars files.

File: docs/docsite/rst/guide_alicloud.rst
Comment (gundalow) [inline docs/docsite/rst/guide_alicloud.rst]: When mentioning option names it would be good to have them formatted as code, rather than just in quotes,

```
... would use ``instance_name`` to create... 
```

File: lib/ansible/module_utils/alicloud_ecs.py
Comment (gundalow) [inline lib/ansible/module_utils/alicloud_ecs.py]: Is `163.com.com` correct?

File: lib/ansible/module_utils/alicloud_ecs.py
Comment (gundalow) [inline lib/ansible/module_utils/alicloud_ecs.py]: Would we want `no_log` here as well?

File: lib/ansible/module_utils/alicloud_ecs.py
Comment (gundalow) [inline lib/ansible/module_utils/alicloud_ecs.py]: You can add this directly into `acs_common_argument_spec`

Eg
`'username': dict(fallback=(env_fallback, ['ANSIBLE_NET_USERNAME'])),`

File: lib/ansible/module_utils/alicloud_ecs.py
Comment (gundalow) [inline lib/ansible/module_utils/alicloud_ecs.py]: This can be removed, just add required =True in `acs_common_argument_spec`

File: lib/ansible/modules/cloud/alicloud/alicloud_instance.py
Comment (gundalow) [inline lib/ansible/modules/cloud/alicloud/alicloud_instance.py]: `.com.com`?

File: lib/ansible/modules/cloud/alicloud/alicloud_instance.py
Comment (gundalow) [inline lib/ansible/modules/cloud/alicloud/alicloud_instance.py]: 2.5

File: lib/ansible/modules/cloud/alicloud/alicloud_instance.py
Comment (gundalow) [inline lib/ansible/modules/cloud/alicloud/alicloud_instance.py]: I wonder if this should include alicloud?

File: lib/ansible/modules/cloud/alicloud/alicloud_instance.py
Comment (gundalow) [inline lib/ansible/modules/cloud/alicloud/alicloud_instance.py]: Please remove this line

File: lib/ansible/modules/cloud/alicloud/alicloud_instance.py
Comment (gundalow) [inline lib/ansible/modules/cloud/alicloud/alicloud_instance.py]: Please remove this line

File: lib/ansible/modules/cloud/alicloud/alicloud_instance.py
Comment (gundalow) [inline lib/ansible/modules/cloud/alicloud/alicloud_instance.py]: Use the following to format his as code:
`C({"key":"value"})`

File: lib/ansible/modules/cloud/alicloud/alicloud_instance.py
Comment (gundalow) [inline lib/ansible/modules/cloud/alicloud/alicloud_instance.py]: I think `state` should be hard coded in the task, not as a variable

File: lib/ansible/modules/cloud/alicloud/alicloud_instance_facts.py
Comment (gundalow) [inline lib/ansible/modules/cloud/alicloud/alicloud_instance_facts.py]: 2.5

File: lib/ansible/modules/cloud/alicloud/alicloud_instance_facts.py
Comment (gundalow) [inline lib/ansible/modules/cloud/alicloud/alicloud_instance_facts.py]: You have 2.6 here and 2.7 in the other module, is that correct?

File: lib/ansible/utils/module_docs_fragments/alicloud.py
Comment (gundalow) [inline lib/ansible/utils/module_docs_fragments/alicloud.py]: `C(ALICLOUD_ACCESS_KEY) `

File: lib/ansible/utils/module_docs_fragments/alicloud.py
Comment (gundalow) [inline lib/ansible/utils/module_docs_fragments/alicloud.py]: 2.6 or 2.7?
Also you `requirements` listed here and the modules.

**Review by gundalow — CHANGES_REQUESTED** [review]

File: lib/ansible/modules/cloud/alicloud/alicloud_instance.py
Comment (xiaozhu36) [reply lib/ansible/modules/cloud/alicloud/alicloud_instance.py]: In order to mark and manage our module better and help users to understand the module, we use 'alicloud' prefix same as aws, azure and other cloud module.

**Comment by xiaozhu36** [comment]
Hi @gundalow 

Thank you very much for your code review. I have improved the modules according to your suggestion. Please check them again, thanks a lot.

In addition,  considering many users want to use our module to manage their infrastructure and deploy their applications, I want to know when our modules can be merged and released.

Looking forward to your reply, thanks.

Regards,
He Guimin

**Comment by gundalow** [comment]
Hi,
I'll need to review the changes. This should get into Ansible 2.5.0, which will be released around the Feb/March 2015 timeframe. We haven't finalised dates yet.

**Comment by xiaozhu36** [comment]
Hi @gundalow 

Looking forward to your good new, thanks a lot.

Regards,
He Guimin

File: docs/docsite/rst/guide_alicloud.rst
Comment (gundalow) [inline docs/docsite/rst/guide_alicloud.rst]: Should link to the vault docs here

File: lib/ansible/module_utils/alicloud_ecs.py
Comment (gundalow) [inline lib/ansible/module_utils/alicloud_ecs.py]: Do we really need so many `aliases`?
How do you refer to this in your existing documentation & API?

File: lib/ansible/modules/cloud/alicloud/alicloud_instance.py
Comment (gundalow) [inline lib/ansible/modules/cloud/alicloud/alicloud_instance.py]: Are there any disks apart from system disks?

If so then this should be named `system_disk_category` and remove the aliases.

File: lib/ansible/modules/cloud/alicloud/alicloud_instance.py
Comment (gundalow) [inline lib/ansible/modules/cloud/alicloud/alicloud_instance.py]: `I(count_tag)`

File: lib/ansible/modules/cloud/alicloud/alicloud_instance.py
Comment (gundalow) [inline lib/ansible/modules/cloud/alicloud/alicloud_instance.py]: `I(count)`

File: lib/ansible/modules/cloud/alicloud/alicloud_instance.py
Comment (gundalow) [inline lib/ansible/modules/cloud/alicloud/alicloud_instance.py:178]: I'd define the (common) `vars:` just once and list the different examples one after each other within the same `tasks:` section.

**Review by gundalow — CHANGES_REQUESTED** [review]

**Comment by xiaozhu36** [comment]
Hi @gundalow

Thank you very much for your code review. I have improved the modules according to your suggestion. Please check them again, thanks a lot.

Regards,
He Guimin

**Review by dharmabumstead — APPROVED** [review]
RST docs OK with some edits. Thanks @xiaozhu36!

**Comment by xiaozhu36** [comment]
@dharmabumstead Thanks a lot.

File: lib/ansible/modules/cloud/alicloud/alicloud_instance_facts.py
Comment (gundalow) [reply lib/ansible/modules/cloud/alicloud/alicloud_instance_facts.py]: You have 2.6 here and 2.7 in the other module, is that correct?

**Comment by dharmabumstead** [comment]
Feel free to rebase and reopen.

**Comment by xiaozhu36** [comment]
Hi @dharmabumstead I have rebased and reopened the PR #36898 . If the PR also needs other defects, please tell me and I will resolve it. I hope the PR can be merged and the module can be released as soon as possible, thanks a lot.

**Pull request closed without merging**
